### PR TITLE
[vtk] fix find lz4 hdf5 pegl

### LINKED ISSUE #15502: VTK - ITK - HDF5 Conflicts and are unusable together
**Describe the bug**

Using both VTK and ITK in a project leads to strange problem around hdf5 when calling FindPackage(ITK ...):

```
CMake Error at E:/Work/vcpkg/installed/x64-windows/share/hdf5/hdf5-targets.cmake:37 (message):
  Some (but not all) targets in this export set were already defined.

  Targets Defined: hdf5::hdf5-shared;hdf5::hdf5_hl-shared

  Targets not yet defined: hdf5::hdf5_cpp-shared;hdf5::hdf5_hl_cpp-shared

Call Stack (most recent call first):
  E:/Work/vcpkg/installed/x64-windows/share/hdf5/hdf5-config.cmake:146 (include)
  E:/Work/vcpkg/installed/x64-windows/share/hdf5/vcpkg-cmake-wrapper.cmake:17 (_find_package)
  E:/Work/vcpkg/scripts/buildsystems/vcpkg.cmake:585 (include)
  E:/Work/vcpkg/installed/x64-windows/share/itk/Modules/ITKHDF5.cmake:15 (find_package)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:76 (include)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:31 (itk_module_load)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:129 (_itk_module_config_recurse)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKConfig.cmake:81 (itk_module_config)
  E:/Work/vcpkg/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  libs/registration/itkRegistrationOp/CMakeLists.txt:3 (find_package)

```

**Environment**
- OS: Windows 10
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29335 for x64

**To Reproduce**
1. use cmake
```cmake
find_package(ITK QUIET COMPONENTS ITKOptimizersv4 ITKTransform REQUIRED)
target_link_libraries(${FWPROJECT_NAME} PUBLIC ITKOptimizersv4 ITKTransform)
 ```

**Additional context**

The error message come from hdf5, but this seems to be linked with VTK, since it provides its own FindHDF5.cmake in which the missing targets (`hdf5::hdf5_cpp-shared;hdf5::hdf5_hl_cpp-shared`) are not present. 

BTW, removing the VTK `FindHDF5.cmake` 'fixes' the problem as cmake use the right hdf5-config.cmake from hdf5 instead of the one from VTK.

Still, I have also issues with lz4 (already reported in https://github.com/microsoft/vcpkg/issues/14978) and also PEGTL:

```
CMake Error at E:/Work/vcpkg/installed/x64-windows/share/vtk/FindPEGTL.cmake:26 (set_target_properties):
  Attempt to promote imported target "taocpp::pegtl" to global scope (by
  setting IMPORTED_GLOBAL) which is not built in this directory.
Call Stack (most recent call first):
  E:/Work/vcpkg/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  E:/Work/vcpkg/installed/x64-windows/share/vtk/VTK-vtk-module-find-packages.cmake:538 (find_package)
  E:/Work/vcpkg/installed/x64-windows/share/vtk/vtk-config.cmake:129 (include)
  E:/Work/vcpkg/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  E:/Work/vcpkg/installed/x64-windows/share/itk/Modules/ITKVtkGlue.cmake:16 (find_package)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:76 (include)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:31 (itk_module_load)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKModuleAPI.cmake:129 (_itk_module_config_recurse)
  E:/Work/vcpkg/installed/x64-windows/share/itk/ITKConfig.cmake:81 (itk_module_config)
  E:/Work/vcpkg/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  libs/registration/itkRegistrationOp/CMakeLists.txt:3 (find_package)
```

Using the same kind of fix as for lz4 should also work...

### PR DESCRIPTION
**Pull request by dweckmann**
**Describe the pull request**

- What does your PR fix? Fixes #15502

Basically it fixes LZ4, HDF5, PEGL import problems, allowing to use VTK / ITK with FindPackage.

### CODE REVIEW

File: ports/vtk/FindHDF5.cmake [hunk 1]
```diff
@@ -1,20 +1,20 @@
 #The original VTK file is overdoing it and still prdouces errors that the target hdf5::hdf5_hl-shared cannot be found in dynamic builds
 find_package(hdf5 QUIET NO_MODULE)
 if(TARGET hdf5::hdf5-shared)
-  set_target_properties(hdf5::hdf5-shared PROPERTIES IMPORTED_GLOBAL TRUE)
-  add_library(hdf5::hdf5 ALIAS hdf5::hdf5-shared)
+  add_library(hdf5::hdf5 INTERFACE IMPORTED)
+  target_link_libraries(hdf5::hdf5 INTERFACE hdf5::hdf5-shared)
```
Comment (Neumann-A) [inline ports/vtk/FindHDF5.cmake:5]: better use 
`set_property(TARGET <target> APPEND PROPERTY INTERFACE_LINK_LIBRARIES <deptarget>)`
everywhere using `target_link_libraries` with imported target is only supported by newer cmake versions.

Comment (dweckmann) [reply ports/vtk/FindHDF5.cmake:5]: Ok will do that !

Comment (Neumann-A) [reply ports/vtk/FindHDF5.cmake:5]: hmm try removing imported here maybe that is enough to make it build vtk.

Comment (dweckmann) [reply ports/vtk/FindHDF5.cmake:5]: Well, I will try.

Comment (dweckmann) [reply ports/vtk/FindHDF5.cmake:5]: unfortunately, it doesn't work. I've tried a lot of things, but I'm out of solution for now.

**Comment by JackBoosY** [comment]
Please ping me if this PR is ready for review.

**Comment by dweckmann** [comment]
> Please ping me if this PR is ready for review.

unfortunately not yet. I didn't manage to fix correctly the initial problem (hdf5 double inclusion when using VTK + ITK)

**Comment by WangZP10969** [comment]
> > Please ping me if this PR is ready for review.
> 
> unfortunately not yet. I didn't manage to fix correctly the initial problem (hdf5 double inclusion when using VTK + ITK)

For users who only use vtk, do you have any way to deal with this problem? I tried your changes, for the modification of lz4, the compilation did not report an error, but for hdf5 pegl did not seem to solve the problem.

**Comment by dweckmann** [comment]
> > > Please ping me if this PR is ready for review.
> > 
> > 
> > unfortunately not yet. I didn't manage to fix correctly the initial problem (hdf5 double inclusion when using VTK + ITK)
> 
> For users who only use vtk, do you have any way to deal with this problem? I tried your changes, for the modification of lz4, the compilation did not report an error, but for hdf5 pegl did not seem to solve the problem.

Well, I removed the hdf5 patch as it breaks VTK build (as you noticed). I've tried many things with no acceptable solutions. I'm trying to remove the FindHDF5.cmake after build as a workaround for now...

If any of you have an idea, I'm taking it.

**Comment by dweckmann** [comment]
I didn't manage to fix the issue with hdf5. So I propose this workaround (simply removing FindHDF5.cmake after build from VTK and use the regular config from HDF5 directly), which seems to be effective for us.

If anyone have a better solution, I'll take it.

**Comment by JackBoosY** [comment]
Since you've closed the modify permission of other mantainers, please do the following steps:
1. update this branch from vcpkg master
2. rebuild vcpkg
3. use command `./vcpkg x-add-version --all`
4. update changes

**Comment by WangZP10969** [comment]
@dweckmann Thank you very much for your contribution！
I tried your changes, I can successfully configure the VTK library.

During the test, a problem was found.
Normally, the library is only searched once, but in my development, the library is searched again in the subdirectory (due to design issues).

For other libraries, there is no problem with repeated search:
find_package(Qt5 REQUIRED Widgets)
find_package(Qt5 REQUIRED Widgets)

For the VTK library, an error message will be reported for repeated searches:
find_package(VTK REQUIRED)
find_package(VTK REQUIRED)

`
Searching for PEGTL
Searching for PEGTL - found target taocpp::pegtl
CMake Error at libwvl/3rdParty/scripts/buildsystems/vcpkg.cmake:489 (_add_library):
  _add_library cannot create imported target "PEGTL::PEGTL" because another
  target with the same name already exists.
Call Stack (most recent call first):
  libwvl/3rdParty/installed/x64-windows/share/vtk/FindPEGTL.cmake:26 (add_library)
  libwvl/3rdParty/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  libwvl/3rdParty/installed/x64-windows/share/vtk/VTK-vtk-module-find-packages.cmake:585 (find_package)
  libwvl/3rdParty/installed/x64-windows/share/vtk/vtk-config.cmake:129 (include)
  libwvl/3rdParty/scripts/buildsystems/vcpkg.cmake:634 (_find_package)
  CMakeLists.txt:14 (find_package)
`
Do you have any suggestions for this situation? If finding the library twice is indeed abnormal behavior, I will try to adjust the project logic to solve it.

**Comment by dweckmann** [comment]
> 
> Do you have any suggestions for this situation? If finding the library twice is indeed abnormal behavior, I will try to adjust the project logic to solve it.

No sorry, VTK cmake files are too complex for my basic knowledge about it. I've tried to tweak a bit this part, but either my project won't build, either VTK won't.

**Comment by Neumann-A** [comment]
>Do you have any suggestions for this situation? If finding the library twice is indeed abnormal behavior, I will try to adjust the project logic to solve it.

All find modules should have a check `if(NOT TARGET <targetname>)` and only create the target if it does not exists. Maybe that if is missing here

File: ports/vtk/pegtl.patch [hunk 2]
```diff
@@ -15,8 +15,8 @@ index 73eee02f7..22d8bc159 100644
 +find_package(PEGTL CONFIG NAMES PEGTL-2)
 +if(TARGET taocpp::pegtl)
```
Comment (Neumann-A) [inline ports/vtk/pegtl.patch:16]:
```suggestion
+if(TARGET taocpp::pegtl AND NOT TARGET PEGTL::PEGTL)

```

File: port_versions/v-/vtk.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "6f325c609afb0c171d6d70c965e041f8c7288bd3",
+      "git-tree": "64df1fbc7267f7406b924efa6a09fae854278455",
```
Comment (JackBoosY) [inline port_versions/v-/vtk.json:4]:
```suggestion
      "git-tree": "d4c073aa113a8b8c4a47db1a58fd834f4913fbc2",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please bump the port version. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).

And do the following steps:
1. Revert changes a bout `port_versions/v-/vtk.json`
2. Rerun command `./vcpkg x-add-version vtk`
3. Update changes

**Comment by strega-nil** [comment]
Same for this one, you'll have to rebase on upstream/master.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
LGTM, @Neumann-A  do you have any other questions?

**Comment by Neumann-A** [comment]
no

File: versions/v-/vtk.json [hunk 4]
```diff
@@ -1,7 +1,12 @@
 {
   "versions": [
     {
-      "git-tree": "6f325c609afb0c171d6d70c965e041f8c7288bd3",
+      "git-tree": "1bc53f9d3cafa836645770d1ee56caaf01bb294d",
+      "version-string": "9.0.1",
+      "port-version": 4
+    },
+    {
+      "git-tree": "64df1fbc7267f7406b924efa6a09fae854278455",
```
Comment (dan-shaw) [inline versions/v-/vtk.json:9]: We should not modify the `git-tree` of the previous version.

Comment (JackBoosY) [reply versions/v-/vtk.json:9]: Because this PR has conflicts.

**Comment by JackBoosY** [comment]
Can you please merge this branch from master and re-generate the version record?

Thanks,

**Comment by dweckmann** [comment]
> Can you please merge this branch from master and re-generate the version record?
> 
> Thanks,

Done !

Thank you for your time.

**Comment by JackBoosY** [comment]
@dweckmann Seems you need to update the version record again.

**Comment by JackBoosY** [comment]
Will rerun the pipeline test after all test finished.

File: versions/v-/vtk.json [hunk 5]
```diff
@@ -1,224 +1,9 @@
-{
-  "versions": [
-    {
-      "git-tree": "88d95daae73bf5c3413bb18188c81f9ea752e418",
```
Comment (BillyONeal) [inline versions/v-/vtk.json:4]: You need to not damage the existing versions database.

**Review by BillyONeal — COMMENTED** [review]
You need to:

1. Merge with origin/master. Take master's versions/v-/vtk.json
2. Bump Port-Version in ports/vtk/CONTROL
3. Commit that
4. `vcpkg x-add-version vtk`
5. Commit that

I tried to do this for you but you didn't allow maintainer edits when creating the pr.

File: versions/v-/vtk.json (hunk 5 above)
Comment (dweckmann) [reply versions/v-/vtk.json]: It should be good now. Thanks for your time !

**Comment by BillyONeal** [comment]
(Rerunning due to replacing the macos vms)

**Comment by JackBoosY** [comment]
Please add add `paraview:x64-windows-static-md=fail` to _scripts\ci.baseline.txt_.

**Comment by Neumann-A** [comment]
> Please add add paraview:x64-windows-static-md=fail to scripts\ci.baseline.txt.
The qt executable run failed, not related to this PR:

@JackBoosY: Why? This looks like the same issue marble had in another PR. Maybe there is some issue with the newer CMake version?

**Comment by JackBoosY** [comment]
@Neumann-A So can you make a new PR to fix that?

**Comment by JackBoosY** [comment]
Can you please merge to master?

**Comment by JackBoosY** [comment]
Ping @dweckmann for response.

**Comment by dweckmann** [comment]
> Ping @dweckmann for response.

Sorry, I don't have time to spend on this anymore as we are focused on other subjects right now. I really wish I could help further, but I cannot make any promise, sorry again. Feel free to close the PR or re-author it.

**Comment by JackBoosY** [comment]
If you want to continue this PR, please reopen.

**Comment by FrankXie05** [comment]
@dweckmann Is still work for this PR? Please reply me

**Comment by dweckmann** [comment]
> @dweckmann Is still work for this PR? Please reply me

@FrankXie05 The last time I checked, only the patch for lz4 in VTK is still relevant. I guess you have to strip all other patch and only keep that.

Feel free to take this pr as basis and reopen.

**Comment by FrankXie05** [comment]
@dweckmann Thanks for your reply, I will submit a PR  to verify it. :)

**Pull request closed without merging**